File: src/jvm/package/src/doc-examples/java/questpdf/docexamples/text/TextStyleExamples.java
package questpdf.docexamples.text;

import com.questpdf.Settings;
import com.questpdf.fluent.Document;
import com.questpdf.helpers.Colors;
import com.questpdf.helpers.FontFeatures;
import com.questpdf.helpers.PageSize;
import com.questpdf.helpers.Placeholders;
import com.questpdf.infrastructure.ImageCompressionQuality;
import com.questpdf.infrastructure.ImageFormat;
import com.questpdf.infrastructure.ImageGenerationSettings;
import org.junit.jupiter.api.Test;
import questpdf.docexamples.DocExample;

public class TextStyleExamples extends DocExample {

    @Test
    public void fontSize() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(1000f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .column(column -> {
                            column.spacing(10f);

                            column.item()
                                .text("This is small text (16pt)")
                                .fontSize(16f);

                            column.item()
                                .text("This is medium text (24pt)")
                                .fontSize(24f);

                            column.item()
                                .text("This is large text (36pt)")
                                .fontSize(36f);
                        });
                });
            })
            .generateImages(index -> output("text-font-size.webp"), settings);
    }

    @Test
    public void fontFamily() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(1000f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .column(column -> {
                            column.spacing(10f);

                            column.item().text("This is text with default font (Lato)");

                            column.item().text("This is text with Times New Roman font")
                                .fontFamily("Times New Roman");

                            column.item().text("This is text with Courier New font")
                                .fontFamily("Courier New");
                        });
                });
            })
            .generateImages(index -> output("text-font-family.webp"), settings);
    }

    @Test
    public void fontColor() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.Best);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(1000f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .text(text -> {
                            text.span("Each pixel consists of three sub-pixels: ");
                            text.span("red").fontColor(Colors.Red.getMedium());
                            text.span(", ");
                            text.span("green").fontColor(Colors.Green.getMedium());
                            text.span(" and ");
                            text.span("blue").fontColor(Colors.Blue.getMedium());
                            text.span(".");
                        });
                });
            })
            .generateImages(index -> output("text-font-color.webp"), settings);
    }

    @Test
    public void backgroundColor() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.Best);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(500f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .text(text -> {
                            text.span("The term ");
                            text.span("algorithm").backgroundColor(Colors.Yellow.getLighten3()).bold();
                            text.span(" refers to a set of rules or steps used to solve a problem.");
                        });
                });
            })
            .generateImages(index -> output("text-font-background.webp"), settings);
    }

    @Test
    public void italic() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.Best);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(500f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .text(text -> {
                            text.span("In this sentence, the word ");
                            text.span("important").italic();
                            text.span(" is emphasized using italics.");
                        });
                });
            })
            .generateImages(index -> output("text-font-italic.webp"), settings);
    }

    @Test
    public void fontWeight() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.Best);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(500f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .text(text -> {
                            text.span("This sentence demonstrates ");
                            text.span("bold").bold();
                            text.span(", ");
                            text.span("normal").normalWeight();
                            text.span(", ");
                            text.span("light").light();
                            text.span(" and ");
                            text.span("thin").thin();
                            text.span(" font weights.");
                        });
                });
            })
            .generateImages(index -> output("text-font-weight.webp"), settings);
    }

    @Test
    public void subscript() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.Best);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(500f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .text(text -> {
                            text.span("H");
                            text.span("2").subscript();
                            text.span("O is the chemical formula for water.");
                        });
                });
            })
            .generateImages(index -> output("text-subscript.webp"), settings);
    }

    @Test
    public void superscript() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.Best);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(1000f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .text(text -> {
                            text.span("E = mc");
                            text.span("2").superscript();
                            text.span(" is the equation of mass-energy equivalence.");
                        });
                });
            })
            .generateImages(index -> output("text-superscript.webp"), settings);
    }

    @Test
    public void lineHeight() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.Best);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(500f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .column(column -> {
                            column.spacing(20f);

                            var lineHeights = new float[] { 0.75f, 1f, 2f };
                            var paragraph = Placeholders.paragraph();

                            for (var lineHeight : lineHeights) {
                                column.item()
                                    .background(Colors.Grey.getLighten3())
                                    .padding(5f)
                                    .text(paragraph)
                                    .fontSize(16f)
                                    .lineHeight(lineHeight);
                            }
                        });
                });
            })
            .generateImages(index -> output("text-line-height.webp"), settings);
    }

    @Test
    public void letterSpacing() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.Best);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(500f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .column(column -> {
                            column.spacing(20f);

                            var letterSpacing = new float[] { -0.08f, 0f, 0.2f };
                            var paragraph = Placeholders.sentence();

                            for (var spacing : letterSpacing) {
                                column.item()
                                    .background(Colors.Grey.getLighten3())
                                    .padding(5f)
                                    .text(paragraph)
                                    .fontSize(18f)
                                    .letterSpacing(spacing);
                            }
                        });
                });
            })
            .generateImages(index -> output("text-letter-spacing.webp"), settings);
    }

    @Test
    public void wordSpacing() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.Best);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(500f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .column(column -> {
                            column.spacing(20f);

                            var wordSpacing = new float[] { -0.2f, 0f, 0.4f };
                            var paragraph = Placeholders.sentence();

                            for (var spacing : wordSpacing) {
                                column.item()
                                    .background(Colors.Grey.getLighten3())
                                    .padding(5f)
                                    .text(paragraph)
                                    .fontSize(16f)
                                    .wordSpacing(spacing);
                            }
                        });
                });
            })
            .generateImages(index -> output("text-word-spacing.webp"), settings);
    }

    @Test
    public void fontFallback() {
        Settings.setUseEnvironmentFonts(false);

        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.Best);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(600f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .text("The Arabic word for programming is البرمجة.")
                        .fontFamily("Lato", "Noto Sans Arabic");
                });
            })
            .generateImages(index -> output("text-font-fallback.webp"), settings);
    }

    @Test
    public void fontFallbackEmoji() {
        Settings.setUseEnvironmentFonts(false);

        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.Best);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(600f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .text("Popular emojis include 😊, 😂, ❤️, 👍, and 😎.")
                        .fontFamily("Lato", "Noto Emoji");
                });
            })
            .generateImages(index -> output("text-font-fallback-emoji.webp"), settings);
    }

    @Test
    public void textFontFeatures() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.Best);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(500f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .row(row -> {
                            row.spacing(25f);

                            row.relativeItem()
                                .background(Colors.Grey.getLighten3())
                                .padding(10f)
                                .column(column -> {
                                    column.item().text("Without ligatures").fontSize(16f);

                                    column.item()
                                        .text("fly and fight")
                                        .fontSize(32f)
                                        .disableFontFeature(FontFeatures.StandardLigatures);
                                });

                            row.relativeItem()
                                .background(Colors.Grey.getLighten3())
                                .padding(10f)
                                .column(column -> {
                                    column.item().text("With ligatures").fontSize(16f);

                                    column.item().text("fly and fight")
                                        .fontSize(32f)
                                        .enableFontFeature(FontFeatures.StandardLigatures);
                                });
                        });
                });
            })
            .generateImages(index -> output("text-font-features.webp"), settings);
    }

    @Test
    public void decorationTypes() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.Best);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(500f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .text(text -> {
                            text.span("There are a couple of available text decorations: ");
                            text.span("underline").underline().fontColor(Colors.Red.getMedium());
                            text.span(", ");
                            text.span("strikethrough").strikethrough().fontColor(Colors.Green.getMedium());
                            text.span(" and ");
                            text.span("overline").overline().fontColor(Colors.Blue.getMedium());
                            text.span(". ");
                        });
                });
            })
            .generateImages(index -> output("text-decoration-types.webp"), settings);
    }

    @Test
    public void decorationStyles() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.Best);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(500f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .text(text -> {
                            text.span("Moreover, the decoration can be ");

                            text.span("solid").fontColor(Colors.Indigo.getMedium()).underline().decorationSolid();
                            text.span(", ");
                            text.span("double").fontColor(Colors.Blue.getMedium()).underline().decorationDouble();
                            text.span(", ");
                            text.span("wavy").fontColor(Colors.LightBlue.getMedium()).underline().decorationWavy();
                            text.span(", ");
                            text.span("dotted").fontColor(Colors.Cyan.getMedium()).underline().decorationDotted();
                            text.span(" or ");
                            text.span("dashed").fontColor(Colors.Green.getMedium())
                                .underline().decorationDashed();
                            text.span(".");
                        });
                });
            })
            .generateImages(index -> output("text-decoration-styles.webp"), settings);
    }

    @Test
    public void decorationsAdvanced() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.Best);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(500f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .text(text -> {
                            text.span("This text contains a ");

                            text.span("seriuos")
                                .underline()
                                .decorationWavy()
                                .decorationColor(Colors.Red.getMedium())
                                .decorationThickness(2f);

                            text.span(" typo.");
                        });
                });
            })
            .generateImages(index -> output("text-decoration-advanced.webp"), settings);
    }
}

File: src/jvm/package/src/doc-examples/kotlin/questpdf/docexamples/text/TextStyleExamples.kt
package questpdf.docexamples.text

import com.questpdf.Settings
import com.questpdf.fluent.Document
import com.questpdf.helpers.Colors
import com.questpdf.helpers.FontFeatures
import com.questpdf.helpers.PageSize
import com.questpdf.helpers.Placeholders
import com.questpdf.infrastructure.ImageCompressionQuality
import com.questpdf.infrastructure.ImageFormat
import com.questpdf.infrastructure.ImageGenerationSettings
import org.junit.jupiter.api.Test
import questpdf.docexamples.DocExample

class TextStyleExamples : DocExample() {

    @Test
    fun fontSize() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(1000f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .column {
                            spacing(10f)

                            item()
                                .text("This is small text (16pt)")
                                .fontSize(16f)

                            item()
                                .text("This is medium text (24pt)")
                                .fontSize(24f)

                            item()
                                .text("This is large text (36pt)")
                                .fontSize(36f)
                        }
                }
            }
            .generateImages({ output("text-font-size.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun fontFamily() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(1000f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .column {
                            spacing(10f)

                            item().text("This is text with default font (Lato)")

                            item().text("This is text with Times New Roman font")
                                .fontFamily("Times New Roman")

                            item().text("This is text with Courier New font")
                                .fontFamily("Courier New")
                        }
                }
            }
            .generateImages({ output("text-font-family.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun fontColor() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(1000f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .text {
                            span("Each pixel consists of three sub-pixels: ")
                            span("red").fontColor(Colors.Red.Medium)
                            span(", ")
                            span("green").fontColor(Colors.Green.Medium)
                            span(" and ")
                            span("blue").fontColor(Colors.Blue.Medium)
                            span(".")
                        }
                }
            }
            .generateImages({ output("text-font-color.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.Best
                rasterDpi = 144
            })
    }

    @Test
    fun backgroundColor() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(500f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .text {
                            span("The term ")
                            span("algorithm").backgroundColor(Colors.Yellow.Lighten3).bold()
                            span(" refers to a set of rules or steps used to solve a problem.")
                        }
                }
            }
            .generateImages({ output("text-font-background.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.Best
                rasterDpi = 144
            })
    }

    @Test
    fun italic() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(500f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .text {
                            span("In this sentence, the word ")
                            span("important").italic()
                            span(" is emphasized using italics.")
                        }
                }
            }
            .generateImages({ output("text-font-italic.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.Best
                rasterDpi = 144
            })
    }

    @Test
    fun fontWeight() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(500f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .text {
                            span("This sentence demonstrates ")
                            span("bold").bold()
                            span(", ")
                            span("normal").normalWeight()
                            span(", ")
                            span("light").light()
                            span(" and ")
                            span("thin").thin()
                            span(" font weights.")
                        }
                }
            }
            .generateImages({ output("text-font-weight.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.Best
                rasterDpi = 144
            })
    }

    @Test
    fun subscript() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(500f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .text {
                            span("H")
                            span("2").subscript()
                            span("O is the chemical formula for water.")
                        }
                }
            }
            .generateImages({ output("text-subscript.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.Best
                rasterDpi = 144
            })
    }

    @Test
    fun superscript() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(1000f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .text {
                            span("E = mc")
                            span("2").superscript()
                            span(" is the equation of mass-energy equivalence.")
                        }
                }
            }
            .generateImages({ output("text-superscript.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.Best
                rasterDpi = 144
            })
    }

    @Test
    fun lineHeight() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(500f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .column {
                            spacing(20f)

                            val lineHeights = arrayOf(0.75f, 1f, 2f)
                            val paragraph = Placeholders.paragraph()

                            for (lineHeight in lineHeights) {
                                item()
                                    .background(Colors.Grey.Lighten3)
                                    .padding(5f)
                                    .text(paragraph)
                                    .fontSize(16f)
                                    .lineHeight(lineHeight)
                            }
                        }
                }
            }
            .generateImages({ output("text-line-height.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.Best
                rasterDpi = 144
            })
    }

    @Test
    fun letterSpacing() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(500f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .column {
                            spacing(20f)

                            val letterSpacing = arrayOf(-0.08f, 0f, 0.2f)
                            val paragraph = Placeholders.sentence()

                            for (spacing in letterSpacing) {
                                item()
                                    .background(Colors.Grey.Lighten3)
                                    .padding(5f)
                                    .text(paragraph)
                                    .fontSize(18f)
                                    .letterSpacing(spacing)
                            }
                        }
                }
            }
            .generateImages({ output("text-letter-spacing.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.Best
                rasterDpi = 144
            })
    }

    @Test
    fun wordSpacing() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(500f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .column {
                            spacing(20f)

                            val wordSpacing = arrayOf(-0.2f, 0f, 0.4f)
                            val paragraph = Placeholders.sentence()

                            for (spacing in wordSpacing) {
                                item()
                                    .background(Colors.Grey.Lighten3)
                                    .padding(5f)
                                    .text(paragraph)
                                    .fontSize(16f)
                                    .wordSpacing(spacing)
                            }
                        }
                }
            }
            .generateImages({ output("text-word-spacing.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.Best
                rasterDpi = 144
            })
    }

    @Test
    fun fontFallback() {
        Settings.useEnvironmentFonts = false

        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(600f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .text("The Arabic word for programming is البرمجة.")
                        .fontFamily("Lato", "Noto Sans Arabic")
                }
            }
            .generateImages({ output("text-font-fallback.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.Best
                rasterDpi = 144
            })
    }

    @Test
    fun fontFallbackEmoji() {
        Settings.useEnvironmentFonts = false

        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(600f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .text("Popular emojis include 😊, 😂, ❤️, 👍, and 😎.")
                        .fontFamily("Lato", "Noto Emoji")
                }
            }
            .generateImages({ output("text-font-fallback-emoji.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.Best
                rasterDpi = 144
            })
    }

    @Test
    fun textFontFeatures() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(500f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .row {
                            spacing(25f)

                            relativeItem()
                                .background(Colors.Grey.Lighten3)
                                .padding(10f)
                                .column {
                                    item().text("Without ligatures").fontSize(16f)

                                    item()
                                        .text("fly and fight")
                                        .fontSize(32f)
                                        .disableFontFeature(FontFeatures.StandardLigatures)
                                }

                            relativeItem()
                                .background(Colors.Grey.Lighten3)
                                .padding(10f)
                                .column {
                                    item().text("With ligatures").fontSize(16f)

                                    item().text("fly and fight")
                                        .fontSize(32f)
                                        .enableFontFeature(FontFeatures.StandardLigatures)
                                }
                        }
                }
            }
            .generateImages({ output("text-font-features.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.Best
                rasterDpi = 144
            })
    }

    @Test
    fun decorationTypes() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(500f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .text {
                            span("There are a couple of available text decorations: ")
                            span("underline").underline().fontColor(Colors.Red.Medium)
                            span(", ")
                            span("strikethrough").strikethrough().fontColor(Colors.Green.Medium)
                            span(" and ")
                            span("overline").overline().fontColor(Colors.Blue.Medium)
                            span(". ")
                        }
                }
            }
            .generateImages({ output("text-decoration-types.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.Best
                rasterDpi = 144
            })
    }

    @Test
    fun decorationStyles() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(500f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .text {
                            span("Moreover, the decoration can be ")

                            span("solid").fontColor(Colors.Indigo.Medium).underline().decorationSolid()
                            span(", ")
                            span("double").fontColor(Colors.Blue.Medium).underline().decorationDouble()
                            span(", ")
                            span("wavy").fontColor(Colors.LightBlue.Medium).underline().decorationWavy()
                            span(", ")
                            span("dotted").fontColor(Colors.Cyan.Medium).underline().decorationDotted()
                            span(" or ")
                            span("dashed").fontColor(Colors.Green.Medium)
                                .underline().decorationDashed()
                            span(".")
                        }
                }
            }
            .generateImages({ output("text-decoration-styles.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.Best
                rasterDpi = 144
            })
    }

    @Test
    fun decorationsAdvanced() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(500f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .text {
                            span("This text contains a ")

                            span("seriuos")
                                .underline()
                                .decorationWavy()
                                .decorationColor(Colors.Red.Medium)
                                .decorationThickness(2f)

                            span(" typo.")
                        }
                }
            }
            .generateImages({ output("text-decoration-advanced.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.Best
                rasterDpi = 144
            })
    }
}
